core/scale/scale_encoder_stream.hpp

```diff
@@ -0,0 +1,227 @@
+      put(v.begin(), v.end());
+      return *this;
+    }
+    /**
+     * @brief scale-encodes common::Blob as sequence of bytes, not collection
+     * @tparam size size of blob
+     * @param v blob to encode
+     * @return reference to stream
+     */
+    template <size_t size>
+    ScaleEncoderStream &operator<<(const common::Blob<size> & v) {
+      put(v.begin(), v.end());
+      return *this;
+    }
+    /**
+     * @brief scale-encodes common::Buffer as collection of bytes
+     * @param v buffer to encode
+     * @return reference to stream
+     */
+    ScaleEncoderStream &operator<<(const Buffer &v) {
+      return *this << v.toVector();
+    }
+
+    ScaleEncoderStream &operator<<(std::string_view sv) {
+      *this << BigInteger(sv.size());
+      return put(sv.begin(), sv.end());
+    }
+
+    /**
+     * @brief scale-encodes uint8_t value as fixed size integer
+     * @param v value to encode
+     * @return reference to stream
+     */
+    ScaleEncoderStream &operator<<(uint8_t v);
+    /**
+     * @brief scale-encodes int8_t value as fixed size integer
+     * @param v value to encode
+     * @return reference to stream
+     */
+    ScaleEncoderStream &operator<<(int8_t v);
+    /**
+     * @brief scale-encodes uint16_t value as fixed size integer
+     * @param v value to encode
+     * @return reference to stream
+     */
+    ScaleEncoderStream &operator<<(uint16_t v);
+    /**
+     * @brief scale-encodes int16_t value as fixed size integer
+     * @param v value to encode
+     * @return reference to stream
+     */
+    ScaleEncoderStream &operator<<(int16_t v);
+    /**
+     * @brief scale-encodes uint32_t value as fixed size integer
+     * @param v value to encode
+     * @return reference to stream
+     */
+    ScaleEncoderStream &operator<<(uint32_t v);
+    /**
+     * @brief scale-encodes int32_t value as fixed size integer
+     * @param v value to encode
+     * @return reference to stream
+     */
+    ScaleEncoderStream &operator<<(int32_t v);
+    /**
+     * @brief scale-encodes uint64_t value as fixed size integer
+     * @param v value to encode
+     * @return reference to stream
+     */
+    ScaleEncoderStream &operator<<(uint64_t v);
+    /**
+     * @brief scale-encodes int64_t value as fixed size integer
+     * @param v value to encode
+     * @return reference to stream
+     */
+    ScaleEncoderStream &operator<<(int64_t v);
+    /**
+     * @brief scale-encodes BigInteger value as compact integer
+     * @param v value to encode
+     * @return reference to stream
+     */
+    ScaleEncoderStream &operator<<(const BigInteger &v);
+    /**
+     * scale-encodes tribool value
+     * @param v value to encode
+     * @return reference to stream
+     */
+    ScaleEncoderStream &operator<<(tribool v);
+    /**
+     * scale-encodes bool value
+     * @param v value to encode
+     * @return reference to stream
+     */
+    ScaleEncoderStream &operator<<(bool v);
```

Comment:
```C++
template <typename T>
ScaleEncoderStream& operator <<(T&& t){
    if constexpr (std::is_integral<T>::value || std::is_same<T, BigInteger>::value){
        return encodeInteger<T>(t);
    } else if constexpr (std::is_same<T, bool>::value){
        return encodeBool(t);
    } else if constexpr (std::is_same<T, tribool>::value){
        return encodeTribool(t);
    } else {
        static_assert(always_false<T>, "no operator<< for this type");
    }
}
```

Reply:
It shouldn't be implemented exactly like this. First, bool is integral type and to differentiate it, we need to check first std::is_same<T, bool> and then std::is_integral<T>;
Second, not every integer should be encoded as BigInteger, it is only for compact encoded numbers.
Third, declaring universal operator is problematic, since it fits for any type.

I'm trying to use SFINAE to declare operator only for integral types, but not successfully yet.

Reply:
Done